astartes/utils/array_type_helpers.py

```diff
@@ -0,0 +1,203 @@
+                )
+            )
+
+    return new_array
+
+
+def panda_handla(X, y, labels):
+    """Helper function to deal with supporting Pandas data types in astartes
+
+    Args:
+        X (Dataframe): Features with column names
+        y (Series): Targets
+        labels (Series): Labels for data
+
+    Returns:
+        dict: Empty if no pandas types, metadata-filled otherwise
+    """
+    output_is_pandas = {}
+    # X can be a dataframe, in which case save the column names
+    if type(X) is pd.DataFrame:
+        output_is_pandas["X"] = {"columns": X.columns, "index": X.index}
+
+    # y can be a series, indicate with True
+    if type(y) is pd.Series:  # this implicitly avoids None types
+        output_is_pandas["y"] = {"index": y.index}
+
+    # labels can also be a series, indicate with True
+    if type(labels) is pd.Series:
+        output_is_pandas["labels"] = {"index": labels.index}
+
+    return output_is_pandas
+
+
+def return_helper(
+    sampler_instance,
+    train_idxs,
+    val_idxs,
+    test_idxs,
+    return_indices,
+    output_is_pandas,
+):
+    """Convenience function to return the requested arrays appropriately.
+
+    Args:
+        sampler_instance (sampler): The fit sampler instance.
+        test_size (float): Fraction of data to use in test.
+        val_size (float): Fraction of data to use in val.
+        train_size (float): Fraction of data to use in train.
+        return_indices (bool): Return indices after the value arrays.
+        output_is_pandas (dict): metadata about casting to pandas.
+
+    Returns:
+        np.array: Either many arrays or indices in arrays.
+
+    Notes:
+        This function copies and pastes a lot of code when it could instead
+        use some loop over (X, y, labels, sampler_instance.get_clusters())
+        but such an implementation is more error prone. This is long and
+        not the prettiest, but it is definitely doing what we want.
+    """
+    # pack all the output into a list and then unpack it in the return statement
+    out = []
+
+    # Feature array
+    X_train = sampler_instance.X[train_idxs]
+    # check if it needs to be converted back to a Dataframe
+    dataframe_metadata = output_is_pandas.get("X", False)
+    if dataframe_metadata:
+        X_train = pd.DataFrame(
+            X_train,
+            columns=dataframe_metadata["columns"],
+            index=dataframe_metadata["index"][train_idxs],
+        )
+    out.append(X_train)
+    # repeat the above process for validation and test
+    if len(val_idxs):
+        X_val = sampler_instance.X[val_idxs]
+        if dataframe_metadata:
+            X_val = pd.DataFrame(
+                X_val,
+                columns=dataframe_metadata["columns"],
+                index=dataframe_metadata["index"][val_idxs],
+            )
+        out.append(X_val)
+    X_test = sampler_instance.X[test_idxs]
+    if dataframe_metadata:
+        X_train = pd.DataFrame(
+            X_train,
```

Comment:
I believe this should be `X_test`? Not sure if this impacts the unittests

Reply:
Actually it looks like the unittests only test `train_X` but not `test_X`. Could we also test `test_X`?